### PR TITLE
CBL-1361 : logging the configs of Listener and Replicator at LogInfo.

### PR DESCRIPTION
**Pull request by jianminzhao**

### CODE REVIEW

File: REST/c4Listener.cc [hunk 1]
```diff
@@ -38,6 +38,40 @@ string C4Listener::URLNameFromPath(slice pathSlice) {
     return Listener::databaseNameFromPath(FilePath(pathSlice, ""));
 }
 
+namespace {
+    std::stringstream& operator<<(std::stringstream& ss, const C4ListenerConfig& config) {
+        ss << "{" << "apis: " << (config.apis == kC4RESTAPI ? "REST" : "Sync") << ", "
+        << "networkInterface: " << std::string(config.networkInterface) << ", "
```
Comment (pasin) [inline REST/c4Listener.cc:44]: if `config.networkInterface` is nullslice, should we put null or "NULL" here?

File: REST/c4Listener.cc [hunk 2]
```diff
@@ -38,6 +38,40 @@ string C4Listener::URLNameFromPath(slice pathSlice) {
...
+            << "certificate: " << (config.tlsConfig->certificate == nullptr ? "NULL" : "***") << ", "
+            << "requireClientCerts: " << config.tlsConfig->requireClientCerts << ", "
+            << "rootClientCerts: " << (config.tlsConfig->rootClientCerts == nullptr ? "NULL" : "***") << ", "
+            << "certAuthCallback: " << (config.tlsConfig->certAuthCallback == nullptr ? "NULL" : "***") << ", "
+            << "tlsCallbackContext: " << (config.tlsConfig->tlsCallbackContext == nullptr ? "NULL" : "***");
+        }
+        ss << "}, "
+        << "httpAuthCallback: " << (config.httpAuthCallback == nullptr ? "NULL" : "***") << ", "
+        << "callbackContext: " << (config.callbackContext == nullptr ? "NULL" : "***") << ", "
+        << "directory: " << std::string(config.directory) << ", ";
```
Comment (pasin) [inline REST/c4Listener.cc:61]: If config.directory is nullslice, I think we should put NULL here.

Comment (jianminzhao) [reply REST/c4Listener.cc:61]: same above

File: REST/c4Listener.cc [hunk 3]
```diff
@@ -38,6 +38,45 @@ string C4Listener::URLNameFromPath(slice pathSlice) {
     return Listener::databaseNameFromPath(FilePath(pathSlice, ""));
 }
 
+namespace {
+    bool isEmptyString(const C4String& c4str) {
+        return !c4str || c4str.size == 0;
+    }
```
Comment (borrrden) [inline REST/c4Listener.cc:44]: Instead of this, why not have a `toString` method that gets rid of the repeated tertiary operator?

Comment (jianminzhao) [reply REST/c4Listener.cc:44]: toString may yield empty string. That's what the previous version does. I think Pasin wants NULL for empty string.
By the way, this is not strict json string value whereby empty string is "", and here empty string is  , really empty.

**Review by borrrden — COMMENTED** [review]
Could someone explain what the value is in adding a bunch of "***" to the log?

File: REST/c4Listener.cc (hunk 3 above)
Comment (borrrden) [reply REST/c4Listener.cc:44]: Right but if you are making a method here anyway, why not have it return the string `"NULL"` here?

**Comment by jianminzhao** [comment]
One purpose "***" can serve, I think, is that the caller did supply something.

**Comment by borrrden** [comment]
Yes, I considered that but have we ever had a case where we weren't sure if the caller supplied us something or not?  I know this logging was requested from outside parties but there was never any details about what exactly they were looking to get out of it.

**Comment by jianminzhao** [comment]
Taking for examples, certAuthCallback and tlsCallbackContext, they are void*, the actual numbers may not be too much useful unless taking together with other bugs where actual pointers are logged, but perhaps the presence of them can lead to different code paths, etc. I am open to what we log. @pasin ?

File: REST/c4Listener.cc (hunk 3 above)
Comment (pasin) [reply REST/c4Listener.cc:44]: I think it's better to separate b/w NULL and empty string. If empty string, the log will print "" and if NULL, the log will print NULL. Also the NULL representation of C4String is { nullptr, 0 }?

Comment (jianminzhao) [reply REST/c4Listener.cc:44]: Let me note: we are not doing json, where a string value is quoted both its binary representation of printed form; we print it for human consumption. So, we will print a C4String in 3 shape:
--------
directory: this/path
directory: ""
directory: NULL
--------
sounds good?

Comment (pasin) [reply REST/c4Listener.cc:44]: **BIG and CLEAR**

Comment (jianminzhao) [reply REST/c4Listener.cc:44]: What is printed:
01:15:08.076671| [Listener]: Listener config: {apis: REST, networkInterface: NULL, port: 0, tlsConfig: {}, httpAuthCallback: NULL, callbackContext: NULL, directory: "", allowCreateDBs: 0, allowDeleteDBs: 0}

**Review by pasin — APPROVED** [review]

**Pull request merged**